add Paddle predictor

Ruminateer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kubeflow/kfserving/blob/master/docs/DEVELOPER_GUIDE.md
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR adds [Paddle](https://github.com/PaddlePaddle/Paddle) support to KFserving. `python/paddleserver` package is implemented. Paddle predictor support is added in v1beta1 API.
Paddle users want an easy way to serve Paddle models, therefore we have [Paddle Serving](https://github.com/PaddlePaddle/Serving). We found that some of the Paddle users also use KFsering. They want to integrate paddle into their KFserving workflow. This PR allows them to use KFserving to serve Paddle models with ease.

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
TODO
```

Ruminateer commented:
The majority of work is done, you can follow `docs/samples/v1beta1/paddle/README.md` for a test run.
Things left to do:
- Complete/Update docs
- Add tests for `predictor_paddle.go`
- Update image address. Now I am using my personal docker repository.
- Add GPU support to paddle server.

Ruminateer commented on config/configmap/inferenceservice.yaml:
```diff
@@ -107,6 +107,14 @@ data:
             "multiModelServer": false
+        },
+        "paddle": {
+            "image": "ruminateer/paddleserver",
+            "defaultImageVersion": "latest",
```
Need to be updated

yuzisun commented:
@Ruminateer Thanks for contributing this! Would you be interested in joining our community meeting on next Wednesday to discuss about this. We are happy to see more framework added to KFServing but the concern would be the long term support for this as our WG currently lacks the knowledge about paddle.

Ruminateer commented:
> @Ruminateer Thanks for contributing this! Would you be interested in joining our community meeting on next Wednesday to discuss about this? We are happy to see more framework added to KFServing but the concern would be the long term support for this as our WG currently lacks the knowledge about paddle.

Cool, how can I join the meeting? Also, would you mind Ti Zhou @tizhou86 also joining the meeting? He is a senior architect at the Paddle team (my supervisor), and he can talk to you about the long term support and other stuffs on behalf of the Paddle team.

yuzisun commented:
> > @Ruminateer Thanks for contributing this! Would you be interested in joining our community meeting on next Wednesday to discuss about this? We are happy to see more framework added to KFServing but the concern would be the long term support for this as our WG currently lacks the knowledge about paddle.
> 
> Cool, how can I join the meeting? Also, would you mind Ti Zhou @tizhou86 also joining the meeting? He is a senior architect at the Paddle team (my supervisor), and he can talk to you about the long term support and other stuffs on behalf of the Paddle team.

ye that would be great! check the [kubeflow calendar](https://calendar.google.com/calendar/u/0/embed?src=kubeflow.org_7l5vnbn8suj2se10sen81d9428@group.calendar.google.com&ctz=America/Los_Angeles), our WG meeting occurs biweekly on Wednesday so the next one is on 5/26 9am PST.

yuzisun commented:
/ok-to-test

tizhou86 commented:
> > > @Ruminateer Thanks for contributing this! Would you be interested in joining our community meeting on next Wednesday to discuss about this? We are happy to see more framework added to KFServing but the concern would be the long term support for this as our WG currently lacks the knowledge about paddle.
> > 
> > 
> > Cool, how can I join the meeting? Also, would you mind Ti Zhou @tizhou86 also joining the meeting? He is a senior architect at the Paddle team (my supervisor), and he can talk to you about the long term support and other stuffs on behalf of the Paddle team.
> 
> ye that would be great! check the [kubeflow calendar](https://calendar.google.com/calendar/u/0/embed?src=kubeflow.org_7l5vnbn8suj2se10sen81d9428@group.calendar.google.com&ctz=America/Los_Angeles), our WG meeting occurs biweekly on Wednesday so the next one is on 5/26 9am PST.

Sounds great! We'd like to join and make a brief introduction of PaddlePaddle.

yuzisun commented on python/paddleserver/paddleserver/model.py:
```diff
@@ -0,0 +1,75 @@
+# Copyright 2019 kubeflow.org.
```
2021

yuzisun commented on docs/samples/v1beta1/custom/paddleserving/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+FROM registry.baidubce.com/paddlepaddle/serving:0.5.0-devel
+
+RUN git clone https://github.com/PaddlePaddle/Serving.git
```
Do we need to clone the repo here?

yuzisun commented on docs/samples/v1beta1/paddle/README.md:
```diff
@@ -0,0 +1,58 @@
+# Predict on an InferenceService using SKLearnServer
```
/SKLearnServer/PaddleServer/

yuzisun commented on pkg/apis/serving/v1beta1/predictor.go:
```diff
@@ -47,6 +48,8 @@ type PredictorSpec struct {
 	PMML *PMMLSpec `json:"pmml,omitempty"`
 	// Spec for LightGBM model server
 	LightGBM *LightGBMSpec `json:"lightgbm,omitempty"`
+	// Spec for Paddle model server
```
Add a link for the Paddle serving?

yuzisun commented on python/paddleserver/README.md:
````diff
@@ -0,0 +1,108 @@
+```
+You can now point to your paddle model directory and use the server to load the model and test for prediction. Models can be on local filesystem, S3 compatible object storage, Azure Blob Storage, or Google Cloud Storage.
+
+## Building your own Scikit-Learn Server Docker Image
````
/Scikit-Learn/Paddle/

yuzisun commented on python/paddleserver/README.md:
````diff
@@ -0,0 +1,108 @@
+
+To push your image to your dockerhub repo,
+```bash
+docker push ${docker_user_name}/sklearnserver:latest
````
/sklearnserver/paddleserver/

yuzisun commented on python/paddleserver/paddleserver/__init__.py:
```diff
@@ -0,0 +1,15 @@
+# Copyright 2019 kubeflow.org.
```
2021

yuzisun commented on python/paddleserver/paddleserver/__main__.py:
```diff
@@ -0,0 +1,31 @@
+# Copyright 2019 kubeflow.org.
```
2021

yuzisun commented on python/paddleserver/paddleserver/test_model.py:
```diff
@@ -0,0 +1,36 @@
+# Copyright 2019 kubeflow.org.
```
2021

yuzisun commented:
@Ruminateer Can you help add the an e2e test like [this example](https://github.com/kubeflow/kfserving/blob/master/test/e2e/predictor/test_sklearn.py) ?

Ruminateer commented:
> @Ruminateer Can you help add the an e2e test like [this example](https://github.com/kubeflow/kfserving/blob/master/test/e2e/predictor/test_sklearn.py) ?

Yes I will be working on that

Ruminateer commented:
Rebased master and generated python client, then somehow e2e tests start to fail with [log](https://argo.kubeflow-testing.com/workflows/kubeflow-test-infra/kubeflow-kfserving-presubmit-e2e-1615-ba89b5f-1936-292b?tab=workflow&nodeId=kubeflow-kfserving-presubmit-e2e-1615-ba89b5f-1936-292b-1348624524&sidePanel=logs%3Akubeflow-kfserving-presubmit-e2e-1615-ba89b5f-1936-292b-1348624524%3Amain). Not sure why yet.

Ruminateer commented:
/retest

Ruminateer commented:
/retest

Ruminateer commented:
/retest

Ruminateer commented:
@yuzisun I added the e2e test, but it is failing with the error
```console
E           RuntimeError: Exception when calling CustomObjectsApi->create_namespaced_custom_object:                 (403)
E           Reason: Forbidden
E           HTTP response headers: HTTPHeaderDict({'Audit-Id': '65111ac9-e031-40dc-a18f-f82083a382c1', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'Date': 'Wed, 02 Jun 2021 10:32:39 GMT', 'Content-Length': '463'})
E           HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"admission webhook \"inferenceservice.kfserving-webhook-server.v1beta1.validator\" denied the request: Exactly one of [SKLearn, XGBoost, Tensorflow, PyTorch, Triton, ONNX, PMML, LightGBM, PodSpec] must be specified in PredictorSpec","reason":"Exactly one of [SKLearn, XGBoost, Tensorflow, PyTorch, Triton, ONNX, PMML, LightGBM, PodSpec] must be specified in PredictorSpec","code":403}
/mnt/test-data-volume/kubeflow-kfserving-presubmit-e2e-1615-747e6ac-0080-b97d/src/kubeflow/kfserving/python/kfserving/kfserving/api/kf_serving_client.py:115: RuntimeError
```
It seems that error occurs at [`func (*InferenceService) ValidateCreate() error`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/inference_service_validation.go#L60),
which calls [`func validateExactlyOneImplementation(Component) error`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/component.go#L179-L184),
where an [error](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/component.go#L211-L214) is returned.

The list in the error message doesn't include Paddle. However, I did add Paddle in [`func (*PredictorSpec) GetPredictorImplementations() []PredictorImplementation`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/predictor.go#L114-L131) and [`type PredictorSpec struct`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/predictor.go#L52). It seems like the test isn't running the exact Go code on my branch. How do I change the configuration of the e2e test to make it works? Or I am I missing something?

yuzisun commented on config/configmap/inferenceservice.yaml:
```diff
@@ -107,6 +107,14 @@ data:
             ],
             "multiModelServer": false
+        },
+        "paddle": {
```
@Ruminateer Please add this section to the [e2e test config overlay](https://github.com/kubeflow/kfserving/blob/master/config/overlays/test/configmap/inferenceservice.yaml).

yuzisun commented:
> @yuzisun I added the e2e test, but it is failing with the error
> 
> ```
> E           RuntimeError: Exception when calling CustomObjectsApi->create_namespaced_custom_object:                 (403)
> E           Reason: Forbidden
> E           HTTP response headers: HTTPHeaderDict({'Audit-Id': '65111ac9-e031-40dc-a18f-f82083a382c1', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'Date': 'Wed, 02 Jun 2021 10:32:39 GMT', 'Content-Length': '463'})
> E           HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"admission webhook \"inferenceservice.kfserving-webhook-server.v1beta1.validator\" denied the request: Exactly one of [SKLearn, XGBoost, Tensorflow, PyTorch, Triton, ONNX, PMML, LightGBM, PodSpec] must be specified in PredictorSpec","reason":"Exactly one of [SKLearn, XGBoost, Tensorflow, PyTorch, Triton, ONNX, PMML, LightGBM, PodSpec] must be specified in PredictorSpec","code":403}
> /mnt/test-data-volume/kubeflow-kfserving-presubmit-e2e-1615-747e6ac-0080-b97d/src/kubeflow/kfserving/python/kfserving/kfserving/api/kf_serving_client.py:115: RuntimeError
> ```
> 
> It seems that error occurs at [`func (*InferenceService) ValidateCreate() error`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/inference_service_validation.go#L60),
> which calls [`func validateExactlyOneImplementation(Component) error`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/component.go#L179-L184),
> where an [error](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/component.go#L211-L214) is returned.
> 
> The list in the error message doesn't include Paddle. However, I did add Paddle in [`func (*PredictorSpec) GetPredictorImplementations() []PredictorImplementation`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/predictor.go#L114-L131) and [`type PredictorSpec struct`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/predictor.go#L52). It seems like the test isn't running the exact Go code on my branch. How do I change the configuration of the e2e test to make it works? Or I am I missing something?

@Ruminateer Yes we need to add paddle to the [e2e test config overlay](https://github.com/kubeflow/kfserving/blob/master/config/overlays/test/configmap/inferenceservice.yaml).

Ruminateer commented:
I added a workflow step in the e2e tests to build the paddle image.
https://github.com/Ruminateer/kfserving/blob/828e7ffc6f7a3cb8a9ec0a6031bc917c1b510356/test/workflows/components/workflows.libsonnet#L408-L413
It seems that Kaniko executor cannot create new image repo on the server. Can you create a repo for the paddleserver? Or can that also be done within this repo? @yuzisun

yuzisun commented:
> I added a workflow step in the e2e tests to build the paddle image.
> https://github.com/Ruminateer/kfserving/blob/828e7ffc6f7a3cb8a9ec0a6031bc917c1b510356/test/workflows/components/workflows.libsonnet#L408-L413
> It seems that Kaniko executor cannot create new image repo on the server. Can you create a repo for the paddleserver? Or can that also be done within this repo? @yuzisun

Ah right, the image needs to be registered. @PatrickXYS are you still able to help add this image?

tizhou86 commented:
Hi @PatrickXYS , please let us know if anything is needed from our side, thanks!

yuzisun commented:
> Hi @PatrickXYS , please let us know if anything is needed from our side, thanks!

@tizhou86 I think @PatrickXYS is on vacation, @theofpa Any thoughts on this?

yuzisun commented:
@Ruminateer @tizhou86 I do not want to block this, we can comment out the paddleserver image build step for now and use the image in e2e test built manually from this PR.

yuzisun commented:
@Ruminateer Please also help squash and clean up the commits, there are 38 commits in the PR.

yuzisun commented:
@Ruminateer We also need to add a github action like [this one](https://github.com/kubeflow/kfserving/blob/master/.github/workflows/lightgbm-docker-publish.yml) for releasing the paddleserver image

Ruminateer commented:
> @Ruminateer We also need to add a github action like [this one](https://github.com/kubeflow/kfserving/blob/master/.github/workflows/lightgbm-docker-publish.yml) for releasing the paddleserver image

Added

> @Ruminateer Please also help squash and clean up the commits, there are 38 commits in the PR.

Do you want me to squash all these into a single commit?🤗

> @Ruminateer @tizhou86 I do not want to block this, we can comment out the paddleserver image build step for now and use the image in e2e test built manually from this PR.

Ruminateer commented:
e2e test is failing because somehow the paddle server image cannot be pulled
rebuilt the image and /retest

Ruminateer commented:
/retest

yuzisun commented:
> e2e test is failing because somehow the paddle server image cannot be pulled
> rebuilt the image

@Ruminateer I think the culprit is [here](https://github.com/kubeflow/kfserving/blob/master/test/scripts/run-e2e-tests.sh#L121) in e2e we replace the image tag with `PULL_BASE_SHA` , I'd have to double check I believe it is the pull request number.

yuzisun commented:
`PULL_BASE_SHA` is actually on the prow job pod which is `9995b0ebde1f53710745acc0bb5257aa140beb13`
```
    env:
    - name: CLOUD_PROVIDER
      value: aws
    - name: AWS_DEFAULT_REGION
      value: us-west-2
    - name: AWS_ACCESS_KEY_ID
      valueFrom:
        secretKeyRef:
          key: AWS_ACCESS_KEY_ID
          name: aws-credentials
    - name: AWS_SECRET_ACCESS_KEY
      valueFrom:
        secretKeyRef:
          key: AWS_SECRET_ACCESS_KEY
          name: aws-credentials
    - name: AWS_EKS_CLUSTER
      value: optional-test-infra-argo
    - name: ARGO_ENDPOINT
      value: https://argo.kubeflow-testing.com/
    - name: ARTIFACTS_S3_BUCKET
      value: aws-kubeflow-jenkins
    - name: BUILD_ID
      value: "1405195074983170048"
    - name: BUILD_NUMBER
      value: "1405195074983170048"
    - name: CI
      value: "true"
    - name: JOB_NAME
      value: kubeflow-kfserving-presubmit
    - name: JOB_SPEC
      value: '{"type":"presubmit","job":"kubeflow-kfserving-presubmit","buildid":"1405195074983170048","prowjobid":"ce72861f-cebc-11eb-b21c-8a0e88738077","refs":{"org":"kubeflow","repo":"kfserving","repo_link":"https://github.com/kubeflow/kfserving","base_ref":"master","base_sha":"9995b0ebde1f53710745acc0bb5257aa140beb13","base_link":"https://github.com/kubeflow/kfserving/commit/9995b0ebde1f53710745acc0bb5257aa140beb13","pulls":[{"number":1615,"author":"Ruminateer","sha":"75a341347a04fadc6d15305afb96dfc43a5c1874","link":"https://github.com/kubeflow/kfserving/pull/1615","commit_link":"https://github.com/kubeflow/kfserving/pull/1615/commits/75a341347a04fadc6d15305afb96dfc43a5c1874","author_link":"https://github.com/Ruminateer"}]}}'
    - name: JOB_TYPE
      value: presubmit
    - name: PROW_JOB_ID
      value: ce72861f-cebc-11eb-b21c-8a0e88738077
    - name: PULL_BASE_REF
      value: master
    - name: PULL_BASE_SHA
      value: 9995b0ebde1f53710745acc0bb5257aa140beb13
    - name: PULL_NUMBER
      value: "1615"
    - name: PULL_PULL_SHA
      value: 75a341347a04fadc6d15305afb96dfc43a5c1874
    - name: PULL_REFS
      value: master:9995b0ebde1f53710745acc0bb5257aa140beb13,1615:75a341347a04fadc6d15305afb96dfc43a5c1874
    - name: REPO_NAME
      value: kfserving
    - name: REPO_OWNER
      value: kubeflow
    image: public.ecr.aws/j1r0q0g6/kubeflow-testing:latest
```

yuzisun commented on config/overlays/test/configmap/inferenceservice.yaml:
```diff
@@ -80,6 +80,14 @@ data:
         },
+        "paddle": {
+            "image": "ruminateer/paddleserver",
+            "defaultImageVersion": "latest",
```
@Ruminateer For now you can pin to a specific tag, in e2e script we replace latest with the PULL_BASED_SHA which is different for each run.

yuzisun commented:
> > @Ruminateer Please also help squash and clean up the commits, there are 38 commits in the PR.
> 
> Do you want me to squash all these into a single commit? I think the PR is anyway squashed and merged by the bot?

when PR is squash merged, it is a single commit but all the comments from previous 38 commits are combined together which makes a really long comment.

Ruminateer commented:
> @Ruminateer I think the culprit is [here](https://github.com/kubeflow/kfserving/blob/master/test/scripts/run-e2e-tests.sh#L121) in e2e we replace the image tag with `PULL_BASE_SHA` , I'd have to double check I believe it is the pull request number.

@yuzisun Yeah, I did miss that. I changed to a specific version, but I am having the following error again.

> @yuzisun I added the e2e test, but it is failing with the error
> 
> ```
> E           RuntimeError: Exception when calling CustomObjectsApi->create_namespaced_custom_object:                 (403)
> E           Reason: Forbidden
> E           HTTP response headers: HTTPHeaderDict({'Audit-Id': '65111ac9-e031-40dc-a18f-f82083a382c1', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'Date': 'Wed, 02 Jun 2021 10:32:39 GMT', 'Content-Length': '463'})
> E           HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"admission webhook \"inferenceservice.kfserving-webhook-server.v1beta1.validator\" denied the request: Exactly one of [SKLearn, XGBoost, Tensorflow, PyTorch, Triton, ONNX, PMML, LightGBM, PodSpec] must be specified in PredictorSpec","reason":"Exactly one of [SKLearn, XGBoost, Tensorflow, PyTorch, Triton, ONNX, PMML, LightGBM, PodSpec] must be specified in PredictorSpec","code":403}
> /mnt/test-data-volume/kubeflow-kfserving-presubmit-e2e-1615-747e6ac-0080-b97d/src/kubeflow/kfserving/python/kfserving/kfserving/api/kf_serving_client.py:115: RuntimeError
> ```
> 
> It seems that error occurs at [`func (*InferenceService) ValidateCreate() error`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/inference_service_validation.go#L60),
> which calls [`func validateExactlyOneImplementation(Component) error`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/component.go#L179-L184),
> where an [error](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/component.go#L211-L214) is returned.
> 
> The list in the error message doesn't include Paddle. However, I did add Paddle in [`func (*PredictorSpec) GetPredictorImplementations() []PredictorImplementation`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/predictor.go#L114-L131) and [`type PredictorSpec struct`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/predictor.go#L52). It seems like the test isn't running the exact Go code on my branch. How do I change the configuration of the e2e test to make it works? Or I am I missing something?

I am confused 😥. When using the `latest` tag, the error is revision missing and the paddle server image cannot be pulled, which as I understand means the object is accepted by the validation webhook in the first place. I [changed the image tag](https://github.com/Ruminateer/kfserving/blob/5642da0b33862f538daa0cd7410de1f0f22485d6/config/overlays/test/configmap/inferenceservice.yaml#L85) and nothing else, and the object is now rejected by the validation webhook 😥.

Also, as I said in the quoted message, I feel like the test isn't running the exact Go code on my branch. Otherwise, at least the error message would include Paddle in the list. Also, I never encountered this problem on my local cluster, so I think this should be related to e2e test. Haven't found the cause 😥.

yuzisun commented:
@Ruminateer that seems odd, the e2e test checks out your PR and builds the controller image, we print the controller log in the argo workflow step `e2e-tests-post-process`, you can add some logging there to debug.

yuzisun commented:
/retest

Ruminateer commented:
@yuzisun [Increased timeout_seconds in e2e test](https://github.com/Ruminateer/kfserving/blob/80bed94df96408b11610a1e03d631aacdf6192eb/test/e2e/predictor/test_paddle.py#L62) and passed e2e test.

Do you want me to squash the commits now? Also, do I keep the [workaround commit](https://github.com/kubeflow/kfserving/pull/1615/commits/5642da0b33862f538daa0cd7410de1f0f22485d6)?

The following error is gone after the retest without any code change. I doubt it's a bug with the e2e test system, but I wasn't able to find it.

> @yuzisun Yeah, I did miss that. I changed to a specific version, but I am having the following error again.
> 
> > @yuzisun I added the e2e test, but it is failing with the error
> > ```
> > E           RuntimeError: Exception when calling CustomObjectsApi->create_namespaced_custom_object:                 (403)
> > E           Reason: Forbidden
> > E           HTTP response headers: HTTPHeaderDict({'Audit-Id': '65111ac9-e031-40dc-a18f-f82083a382c1', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'Date': 'Wed, 02 Jun 2021 10:32:39 GMT', 'Content-Length': '463'})
> > E           HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"admission webhook \"inferenceservice.kfserving-webhook-server.v1beta1.validator\" denied the request: Exactly one of [SKLearn, XGBoost, Tensorflow, PyTorch, Triton, ONNX, PMML, LightGBM, PodSpec] must be specified in PredictorSpec","reason":"Exactly one of [SKLearn, XGBoost, Tensorflow, PyTorch, Triton, ONNX, PMML, LightGBM, PodSpec] must be specified in PredictorSpec","code":403}
> > /mnt/test-data-volume/kubeflow-kfserving-presubmit-e2e-1615-747e6ac-0080-b97d/src/kubeflow/kfserving/python/kfserving/kfserving/api/kf_serving_client.py:115: RuntimeError
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It seems that error occurs at [`func (*InferenceService) ValidateCreate() error`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/inference_service_validation.go#L60),
> > which calls [`func validateExactlyOneImplementation(Component) error`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/component.go#L179-L184),
> > where an [error](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/component.go#L211-L214) is returned.
> > The list in the error message doesn't include Paddle. However, I did add Paddle in [`func (*PredictorSpec) GetPredictorImplementations() []PredictorImplementation`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/predictor.go#L114-L131) and [`type PredictorSpec struct`](https://github.com/Ruminateer/kfserving/blob/747e6aca78001667ef31cfc8fda0a54c541037f5/pkg/apis/serving/v1beta1/predictor.go#L52). It seems like the test isn't running the exact Go code on my branch. How do I change the configuration of the e2e test to make it works? Or I am I missing something?
> 
> I am confused 😥. When using the `latest` tag, the error is revision missing and the paddle server image cannot be pulled, which as I understand means the object is accepted by the validation webhook in the first place. I [changed the image tag](https://github.com/Ruminateer/kfserving/blob/5642da0b33862f538daa0cd7410de1f0f22485d6/config/overlays/test/configmap/inferenceservice.yaml#L85) and nothing else, and the object is now rejected by the validation webhook 😥.
> 
> Also, as I said in the quoted message, I feel like the test isn't running the exact Go code on my branch. Otherwise, at least the error message would include Paddle in the list. Also, I never encountered this problem on my local cluster, so I think this should be related to e2e test. Haven't found the cause 😥.

yuzisun commented:
> @yuzisun [Increased timeout_seconds in e2e test](https://github.com/Ruminateer/kfserving/blob/80bed94df96408b11610a1e03d631aacdf6192eb/test/e2e/predictor/test_paddle.py#L62) and passed e2e test.

@Ruminateer Is that a big model which takes long to start? it is also possible that we have limited resource in the e2e test so need to wait for other tests to release the resources, the cpu resource can be tuned down to fractional cpu like 100m.

> Do you want me to squash the commits now? Also, do I keep the [workaround commit](https://github.com/kubeflow/kfserving/pull/1615/commits/5642da0b33862f538daa0cd7410de1f0f22485d6)?
Yes we can squash the commits now, l am ok to leave the workaround if you can create an issue reminding us changing it later.

yuzisun commented:
@Ruminateer this model is quite big which can bloat up the repo size, do we really need to commit this ?

python/paddleserver/paddleserver/example_models/resnet50/inference.pdmodel

Ruminateer commented:
> @Ruminateer this model is quite big which can bloat up the repo size, do we really need to commit this ?
> 
> python/paddleserver/paddleserver/example_models/resnet50/inference.pdmodel

No, I can change to smaller models, I will change this today.

Ruminateer commented:
used a smaller model

Ruminateer commented:
/retest

Ruminateer commented:
/retest

yuzisun commented on docs/samples/v1beta1/paddle/README.md:
```diff
@@ -0,0 +1,58 @@
+# Predict on an InferenceService using PaddleServer
+In this example, we use a trained paddle resnet50 model to classify images by running an inference service with TorchServe predictor.
```
/TorchServe/Paddle/

yuzisun commented on pkg/apis/serving/v1beta1/predictor_paddle.go:
```diff
@@ -0,0 +1,87 @@
+		fmt.Sprintf("%s=%s", constants.ArgumentHttpPort, constants.InferenceServiceDefaultHttpPort),
+	}
+
+	// TODO: not sure if constants.ArgumentWorkers is correctly set
```
Can remove the comment, this look good to me

yuzisun commented on python/paddleserver/paddleserver/model.py:
```diff
@@ -0,0 +1,75 @@
+        self.predictor = inference.create_predictor(config)
+
+        # TODO: add support for multiple input_names/output_names
+        # are reusing these tensors multi-thread safe? or does it need to be?
```
remember in multiprocessing mode the load is called before the process forking

Ruminateer replied:
While multithread safety isn't guaranteed here, it seems multithreading isn't happening here, but multiprocessing, which shouldn't be a problem.

yuzisun commented on pkg/apis/serving/v1beta1/predictor_paddle.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+func (p *PaddleServerSpec) Validate() error {
+	// TODO: add GPU support
```
We may want to reject if GPU resource is specified?

yuzisun commented:
@Ruminateer A few last small comments, otherwise we are good to go! Can you also help add a OWNERS file under python/paddleserver?

Ruminateer commented:
> Can you also help add a OWNERS file under python/paddleserver?

I am not sure how the file works. Do I just copy https://github.com/kubeflow/kfserving/blob/master/python/kfserving/OWNERS?

yuzisun commented:
> > Can you also help add a OWNERS file under python/paddleserver?
> 
> I am not sure how the file works. Do I just copy https://github.com/kubeflow/kfserving/blob/master/python/kfserving/OWNERS?

Like that format but change to your name, at least two persons are suggested.

Ruminateer commented:
> Like that format but change to your name, at least two persons are suggested.

Added. Seems I need to be a member or a collaborator.

yuzisun commented:
> > Like that format but change to your name, at least two persons are suggested.
> 
> Added. Seems I need to be a member or a collaborator.

yes, @Ruminateer can you send a PR like [this one](https://github.com/kubeflow/internal-acls/pull/433)

yuzisun commented:
Awesome work @Ruminateer !

/lgtm
/approve
